charon-ml/src/Expressions.ml

```diff
@@ -50,16 +40,29 @@ type assumed_fun_id =
 
           We introduce this when desugaring the [ArrayRepeat] rvalue.
        *)
-  | SliceIndexShared
-      (** Converted from [ProjectionElem::Index].
-
-          Signature: `fn<T>(&[T], usize) -> &T`
+  | Index of builtin_index_op
+      (** Converted from indexing `ProjectionElem`s. The signature depends on the parameters. It
+          could look like:
+          - `fn ArrayIndexShared<T,N>(&[T;N], usize) -> &T`
+          - `fn SliceIndexShared<T>(&[T], usize) -> &T`
+          - `fn ArraySubSliceShared<T,N>(&[T;N], usize, usize) -> &[T]`
+          - `fn SliceSubSliceMut<T>(&mut [T], usize, usize) -> &mut [T]`
+          - etc
        *)
-  | SliceIndexMut
-      (** Converted from [ProjectionElem::Index].
 
-          Signature: `fn<T>(&mut [T], usize) -> &mut T`
-       *)
+(** One of 8 built-in indexing operations. *)
+and builtin_index_op = {
+  is_array : bool;  (** Whether this is a slice or array. *)
```

Comment:
It would make sense to have a `kind = Array | Slice` here

Reply:
It would, but I'm finding the struct field names explicit enough